File: 0006-ZigzagConversion/ZigzagConversion.java
public class ZigzagConversion {
    public class Solution {
        public String convert(String s, int numRows) {
            if(numRows == 1 || s.length() <= numRows) return s;
            StringBuilder[] lines = new StringBuilder[numRows];
            StringBuilder result = new StringBuilder();
            int cycleLength = 2 * (numRows - 1);
            for(int row = 0; row < numRows; row++){
                lines[row] = new StringBuilder();
                for(int i = 0; row + i < s.length() ; i += cycleLength){
                    lines[row].append(s.charAt(row + i));
                    if(row != 0 && row != numRows - 1 && i + cycleLength - row < s.length()){
                        lines[row].append(s.charAt(i + cycleLength - row));
                    }
                }
                result.append(lines[row]);
            }
            return result.toString();
        }
    }
}

File: 0006-ZigzagConversion/ZigzagConversion.cpp
#include <string>
#include <vector>
using namespace std;

class Solution{
    public:
        string convert(string s, int numRows) {
        if (numRows == 1 || (int)s.size() <= numRows) return s;

        vector<string> rows(numRows);
        int currentRow = 0;
        bool goingDown = false;

        for (char c : s) {
            rows[currentRow].push_back(c);
            if (currentRow == 0 || currentRow == numRows - 1) {
                goingDown = !goingDown;
            }
            currentRow += goingDown ? 1 : -1;
        }

        string result;
        for (auto &line : rows) {
            result += line;
        }
        return result;
    }
};
